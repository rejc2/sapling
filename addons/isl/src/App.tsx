/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {RepositoryError} from './types';
import type {ReactNode} from 'react';
import type {Writable} from 'shared/typeUtils';

import {CommandHistoryAndProgress} from './CommandHistoryAndProgress';
import {CommitInfoSidebar} from './CommitInfoView/CommitInfoView';
import {CommitTreeList} from './CommitTreeList';
import {ComparisonViewModal} from './ComparisonView/ComparisonViewModal';
import {CwdSelections, availableCwds} from './CwdSelector';
import {Drawers} from './Drawers';
import {EmptyState} from './EmptyState';
import {ErrorBoundary, ErrorNotice} from './ErrorNotice';
import {ISLCommandContext, useCommand} from './ISLShortcuts';
import {Internal} from './Internal';
import {SuspenseBoundary} from './SuspenseBoundary';
import {TopBar} from './TopBar';
import {TopLevelAlerts} from './TopLevelAlert';
import {TopLevelErrors} from './TopLevelErrors';
import {TopLevelToast} from './TopLevelToast';
import {ViewportOverlayRoot} from './ViewportOverlay';
import {tracker} from './analytics';
import {enableReactTools, enableReduxTools} from './atoms/debugToolAtoms';
import {islDrawerState} from './drawerState';
import {GettingStartedModal} from './gettingStarted/GettingStartedModal';
import {I18nSupport, t, T} from './i18n';
import {setJotaiStore} from './jotaiUtils';
import platform from './platform';
import {DEFAULT_RESET_CSS} from './resetStyle';
import {useMainContentWidth, zoomUISettingAtom} from './responsive';
import {repositoryInfo} from './serverAPIState';
import {themeState} from './theme';
import {light} from './tokens.stylex';
import {ModalContainer} from './useModal';
import {usePromise} from './usePromise';
import {isDev, isTest} from './utils';
import * as stylex from '@stylexjs/stylex';
import {VSCodeButton} from '@vscode/webview-ui-toolkit/react';
import {Provider, atom, useAtomValue, useSetAtom, useStore} from 'jotai';
import React, {useMemo} from 'react';
import {ContextMenus} from 'shared/ContextMenu';
import {Icon} from 'shared/Icon';
import {useThrottledEffect} from 'shared/hooks';

import './index.css';

export default function App() {
  return (
    <React.StrictMode>
      <ResetStyle />
      <I18nSupport>
        <MaybeWithJotaiRoot>
          <ISLRoot>
            <ISLCommandContext>
              <ErrorBoundary>
                <NullStateOrDrawers />
                <ViewportOverlayRoot />
                <GettingStartedModal />
                <ComparisonViewModal />
                <ModalContainer />
                <ContextMenus />
                <TopLevelToast />
              </ErrorBoundary>
            </ISLCommandContext>
          </ISLRoot>
        </MaybeWithJotaiRoot>
      </I18nSupport>
    </React.StrictMode>
  );
}

function MaybeWithJotaiRoot({children}: {children: JSX.Element}) {
  if (isTest) {
    // Use a new store when re-mounting so each test (that calls `render(<App />)`)
    // starts with a clean state.
    return (
      <Provider>
        <AccessJotaiRoot />
        {children}
      </Provider>
    );
  } else if (isDev) {
    return <MaybeJotaiDebugTools>{children}</MaybeJotaiDebugTools>;
  } else {
    // Such scoped Provider or store complexity is not needed outside tests or dev.
    return children;
  }
}

const jotaiDevtools = import('./third-party/jotai-devtools/utils');

function MaybeJotaiDebugTools({children}: {children: JSX.Element}) {
  const enabledRedux = useAtomValue(enableReduxTools);
  const enabledReact = useAtomValue(enableReactTools);
  return enabledRedux || enabledReact ? (
    <SuspenseBoundary>
      {enabledRedux ? <AtomsDevtools>{children}</AtomsDevtools> : children}
      {enabledReact && <DebugAtoms />}
    </SuspenseBoundary>
  ) : (
    children
  );
}

function AtomsDevtools({children}: {children: JSX.Element}) {
  const {useAtomsDevtools} = usePromise(jotaiDevtools);
  useAtomsDevtools('jotai');
  return children;
}

function DebugAtoms() {
  const {useAtomsDebugValue} = usePromise(jotaiDevtools);
  useAtomsDebugValue();
  return null;
}

function AccessJotaiRoot() {
  const store = useStore();
  setJotaiStore(store);
  return null;
}

function ResetStyle() {
  const resetCSS = platform.theme?.resetCSS ?? DEFAULT_RESET_CSS;
  return resetCSS.length > 0 ? <style>{resetCSS}</style> : null;
}

function ISLRoot({children}: {children: ReactNode}) {
  const theme = useAtomValue(themeState);
  useAtomValue(zoomUISettingAtom);
  const props = stylex.props(theme === 'light' && light);
  // stylex would overwrite className
  (props as Writable<typeof props>).className += ` isl-root ${theme}-theme`;
  return (
    <div onDragEnter={handleDragAndDrop} onDragOver={handleDragAndDrop}>
      <div {...props}>{children}</div>
    </div>
  );
}

function handleDragAndDrop(e: React.DragEvent<HTMLDivElement>) {
  // VS Code tries to capture drag & drop events to open files. But if you're dragging
  // on ISL, you probably want to do an ImageUpload. Prevent this event from propagating to vscode.
  if (e.dataTransfer?.types?.some(t => t === 'Files')) {
    e.stopPropagation();
    e.preventDefault();
    e.dataTransfer.dropEffect = 'copy';
  }
}

function NullStateOrDrawers() {
  const repoInfo = useAtomValue(repositoryInfo);
  if (repoInfo != null && repoInfo.type !== 'success') {
    return <ISLNullState repoError={repoInfo} />;
  }
  return <ISLDrawers />;
}

function ISLDrawers() {
  const setDrawerState = useSetAtom(islDrawerState);
  useCommand('ToggleSidebar', () => {
    setDrawerState(state => ({
      ...state,
      right: {...state.right, collapsed: !state.right.collapsed},
    }));
  });

  return (
    <Drawers
      rightLabel={
        <>
          <Icon icon="edit" />
          <T>Commit Info</T>
        </>
      }
      right={<CommitInfoSidebar />}
      errorBoundary={ErrorBoundary}>
      <MainContent />
      <CommandHistoryAndProgress />
    </Drawers>
  );
}

function MainContent() {
  const ref = useMainContentWidth();
  return (
    <div className="main-content-area" ref={ref}>
      <TopBar />
      <TopLevelErrors />
      <TopLevelAlerts />
      <CommitTreeList />
    </div>
  );
}

function ISLNullState({repoError}: {repoError: RepositoryError}) {
  const emptyCwds = useAtomValue(useMemo(() => atom(get => get(availableCwds).length === 0), []));
  useThrottledEffect(
    () => {
      if (repoError != null) {
        switch (repoError.type) {
          case 'cwdNotARepository':
            tracker.track('UIEmptyState', {extras: {cwd: repoError.cwd}, errorName: 'InvalidCwd'});
            break;
          case 'invalidCommand':
            tracker.track('UIEmptyState', {
              extras: {command: repoError.command},
              errorName: 'InvalidCommand',
            });
            break;
          case 'unknownError':
            tracker.error('UIEmptyState', 'RepositoryError', repoError.error);
            break;
        }
      }
    },
    1_000,
    [repoError],
  );
  let content;
  if (repoError != null) {
    if (repoError.type === 'cwdNotARepository') {
      if (platform.platformName === 'vscode' && emptyCwds) {
        content = (
          <>
            <EmptyState>
              <div>
                <T>No folder opened</T>
              </div>
              <p>
                <T>Open a folder to get started.</T>
              </p>
            </EmptyState>
          </>
        );
      } else {
        content = (
          <>
            <EmptyState>
              <div>
                <T>Not a valid repository</T>
              </div>
              <p>
                <T replace={{$cwd: <code>{repoError.cwd}</code>}}>
                  $cwd is not a valid Sapling repository. Clone or init a repository to use ISL.
                </T>
              </p>
            </EmptyState>
            <CwdSelections dismiss={() => null} />
          </>
        );
      }
    } else if (repoError.type === 'cwdDoesNotExist') {
      content = (
        <ErrorNotice
          title={
            <T replace={{$cwd: repoError.cwd}}>
              cwd $cwd does not exist. Make sure the folder exists.
            </T>
          }
          error={
            new Error(
              t('$cwd not found', {
                replace: {$cwd: repoError.cwd},
              }),
            )
          }
          buttons={[
            <VSCodeButton
              key="help-button"
              appearance="secondary"
              onClick={e => {
                platform.openExternalLink('https://sapling-scm.com/docs/introduction/installation');
                e.preventDefault();
                e.stopPropagation();
              }}>
              <T>See installation docs</T>
            </VSCodeButton>,
          ]}
        />
      );
    } else if (repoError.type === 'invalidCommand') {
      if (Internal.InvalidSlCommand) {
        content = <Internal.InvalidSlCommand repoError={repoError} />;
      } else {
        content = (
          <ErrorNotice
            startExpanded
            title={<T>Invalid Sapling command. Is Sapling installed correctly?</T>}
            description={
              <T replace={{$cmd: repoError.command}}>Command "$cmd" was not found in PATH</T>
            }
            details={<T replace={{$path: repoError.path ?? '(no path found)'}}>PATH: $path'</T>}
            buttons={[
              <VSCodeButton
                key="help-button"
                appearance="secondary"
                onClick={e => {
                  platform.openExternalLink(
                    'https://sapling-scm.com/docs/introduction/installation',
                  );
                  e.preventDefault();
                  e.stopPropagation();
                }}>
                <T>See installation docs</T>
              </VSCodeButton>,
            ]}
          />
        );
      }
    } else {
      content = <ErrorNotice title={<T>Something went wrong</T>} error={repoError.error} />;
    }
  }

  return <div className="empty-app-state">{content}</div>;
}
