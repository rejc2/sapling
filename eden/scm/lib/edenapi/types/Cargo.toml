# @generated by autocargo

[package]
name = "edenapi_types"
version = "0.1.0"
edition = "2021"

[[test]]
name = "macro_test"
path = "test/basic_test.rs"

[dependencies]
anyhow = "=1.0.72"
bytes = { version = "1.1", features = ["serde"] }
dag-types = { version = "0.1.0", path = "../../dag/dag-types", features = ["for-tests"] }
paste = "1.0.14"
quickcheck = "1.0"
quickcheck_arbitrary_derive = { version = "0.1.0", git = "https://github.com/facebookexperimental/rust-shed.git", branch = "main" }
revisionstore_types = { version = "0.1.0", path = "../../revisionstore/types" }
serde = { version = "1.0.185", features = ["derive", "rc"] }
serde_bytes = "0.11"
serde_derive = "1.0.185"
thiserror = "1.0.43"
type_macros = { version = "0.1.0", path = "proc_macros" }
types = { version = "0.1.0", path = "../../types" }

[dev-dependencies]
insta_ext = { version = "0.1.0", path = "../../insta_ext" }
quickcheck_macros = "1.0"
serde_json = { version = "1.0.100", features = ["float_roundtrip", "unbounded_depth"] }

[features]
default = ["for-tests"]
for-tests = []
