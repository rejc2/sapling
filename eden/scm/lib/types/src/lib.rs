/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Common types used by sibling crates

pub mod errors;
pub mod fetch_mode;
pub mod hash;
pub mod hgid;
pub mod key;
pub mod mutation;
pub mod node;
pub mod nodeinfo;
pub mod parents;
pub mod path;
pub mod repo;
pub mod serde_with;
pub mod sha;
pub mod tree;
pub mod workingcopy_client;

pub use crate::hgid::HgId;
pub use crate::key::Key;
pub use crate::node::Node;
pub use crate::nodeinfo::NodeInfo;
pub use crate::parents::Parents;
pub use crate::path::PathComponent;
pub use crate::path::PathComponentBuf;
pub use crate::path::RepoPath;
pub use crate::path::RepoPathBuf;
pub use crate::sha::Sha256;

pub type Id20 = HgId;

#[cfg(any(test, feature = "for-tests"))]
pub mod testutil;
