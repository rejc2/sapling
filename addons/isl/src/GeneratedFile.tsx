/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {RepoRelativePath} from './types';

import serverAPI from './ClientToServerAPI';
import {t} from './i18n';
import {GeneratedStatus} from './types';
import {useMemo} from 'react';
import {DefaultValue, atom, useRecoilValue} from 'recoil';
import {LRU} from 'shared/LRU';

export const genereatedFileCache = new LRU<RepoRelativePath, GeneratedStatus>(1000);

/** To avoid sending multiple redundant fetch requests, also save which files are being fetched right now */
const currentlyFetching = new Set<RepoRelativePath>();

/**
 * Generated files are cached in `generatedFileCache` LRU.
 * This is not part of a recoil atom so it can be accessed anywhere.
 * In order to allow recoil to rerender dependencies when we update file statuses,
 * store a generation index in recoil.
 * This state should generally be used through useGeneratedFileStatus helpers.
 */
const generatedFileGeneration = atom<number>({
  key: 'generatedFileGeneration',
  default: 0,
  effects: [
    ({setSelf}) => {
      const disposable = serverAPI.onMessageOfType('fetchedGeneratedStatuses', event => {
        for (const [path, status] of Object.entries(event.results)) {
          genereatedFileCache.set(path, status);
          currentlyFetching.delete(path);
        }
        setSelf(old => (old instanceof DefaultValue ? 1 : old + 1));
      });
      return () => disposable.dispose();
    },
  ],
});

export function useGeneratedFileStatus(path: RepoRelativePath): GeneratedStatus {
  useRecoilValue(generatedFileGeneration); // update if we get new statuses
  const found = genereatedFileCache.get(path);
  if (found == null) {
    fetchMissingGeneratedFileStatuses([path]);
    return GeneratedStatus.Manual;
  }
  return found;
}
export function useGeneratedFileStatuses(
  paths: Array<RepoRelativePath>,
): Record<RepoRelativePath, GeneratedStatus> {
  const generation = useRecoilValue(generatedFileGeneration); // update if we get new statuses

  fetchMissingGeneratedFileStatuses(paths);

  return useMemo(() => {
    return Object.fromEntries(
      paths.map(path => [path, genereatedFileCache.get(path) ?? GeneratedStatus.Manual]),
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [generation, paths]);
}

/**
 * Hint that this set of files are being used, any files missing from the generated files cache
 * should be checked on the server.
 * No-op if all files already in the cache.
 */
export function fetchMissingGeneratedFileStatuses(files: Array<RepoRelativePath>) {
  const notCached = files.filter(
    file => genereatedFileCache.get(file) == null && !currentlyFetching.has(file),
  );
  if (notCached.length > 0) {
    for (const file of notCached) {
      currentlyFetching.add(file);
    }
    serverAPI.postMessage({
      type: 'fetchGeneratedStatuses',
      paths: notCached,
    });
  }
}

export function generatedStatusToLabel(status: GeneratedStatus | undefined): string {
  if (status === GeneratedStatus.Generated) {
    return 'generated';
  } else if (status === GeneratedStatus.PartiallyGenerated) {
    return 'partial';
  } else {
    return 'manual';
  }
}

export function generatedStatusDescription(
  status: GeneratedStatus | undefined,
): string | undefined {
  if (status === GeneratedStatus.Generated) {
    return t('This file is generated');
  } else if (status === GeneratedStatus.PartiallyGenerated) {
    return t('This file is partially generated');
  } else {
    return undefined;
  }
}
