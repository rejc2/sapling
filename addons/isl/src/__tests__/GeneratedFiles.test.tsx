/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import App from '../App';
import {genereatedFileCache} from '../GeneratedFile';
import {ignoreRTL} from '../testQueries';
import {
  expectMessageSentToServer,
  simulateCommits,
  COMMIT,
  simulateUncommittedChangedFiles,
  closeCommitInfoSidebar,
  simulateRepoConnected,
  resetTestMessages,
  simulateMessageFromServer,
} from '../testUtils';
import {GeneratedStatus} from '../types';
import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import {act} from 'react-dom/test-utils';

jest.mock('../MessageBus');

describe('Generated Files', () => {
  beforeEach(() => {
    resetTestMessages();
    render(<App />);
    genereatedFileCache.clear();
    act(() => {
      simulateRepoConnected();
      closeCommitInfoSidebar();
      expectMessageSentToServer({
        type: 'subscribe',
        kind: 'smartlogCommits',
        subscriptionID: expect.anything(),
      });
      simulateCommits({
        value: [
          COMMIT('1', 'some public base', '0', {phase: 'public'}),
          COMMIT('a', 'My Commit', '1'),
          COMMIT('b', 'Another Commit', 'a', {isHead: true}),
        ],
      });
      expectMessageSentToServer({
        type: 'subscribe',
        kind: 'uncommittedChanges',
        subscriptionID: expect.anything(),
      });
    });
  });

  async function simulateGeneratedFiles(num: number) {
    const files = new Array(num).fill(null).map((_, i) => `file_${zeroPad(i)}.txt`);
    act(() => {
      simulateUncommittedChangedFiles({
        value: files.map(path => ({
          path,
          status: 'M',
        })),
      });
    });
    await waitFor(() => {
      expectMessageSentToServer({
        type: 'fetchGeneratedStatuses',
        paths: expect.anything(),
      });
    });
    act(() => {
      simulateMessageFromServer({
        type: 'fetchedGeneratedStatuses',
        results: Object.fromEntries(
          files.map((path, i) => [
            path,
            i % 3 === 0
              ? GeneratedStatus.Generated
              : i % 3 === 1
              ? GeneratedStatus.PartiallyGenerated
              : GeneratedStatus.Manual,
          ]),
        ),
      });
    });
  }

  function zeroPad(n: number): string {
    return ('000' + n.toString()).slice(-3);
  }

  it('fetches generated files for uncommitted changes', async () => {
    await simulateGeneratedFiles(5);
    expectMessageSentToServer({
      type: 'fetchGeneratedStatuses',
      paths: ['file_000.txt', 'file_001.txt', 'file_002.txt', 'file_003.txt', 'file_004.txt'],
    });
  });

  it('Shows generated files in their own sections', async () => {
    await simulateGeneratedFiles(10);

    expect(screen.getByText(ignoreRTL('file_002.txt'))).toBeInTheDocument();
    expect(screen.getByText(ignoreRTL('file_005.txt'))).toBeInTheDocument();
    expect(screen.getByText(ignoreRTL('file_008.txt'))).toBeInTheDocument();
    expect(screen.getByText('Generated Files')).toBeInTheDocument();
    expect(screen.getByText('Partially Generated Files')).toBeInTheDocument();
  });

  function goToNextPage() {
    fireEvent.click(screen.getByTestId('changed-files-next-page'));
  }

  function expectHasPartiallyGeneratedFiles() {
    expect(screen.queryByText('Partially Generated Files')).toBeInTheDocument();
  }
  function expectHasGeneratedFiles() {
    expect(screen.queryByText('Generated Files')).toBeInTheDocument();
  }
  function expectNOTHasGeneratedFiles() {
    expect(screen.queryByText('Generated Files')).not.toBeInTheDocument();
  }

  function getChangedFiles() {
    const found = [...document.querySelectorAll('.changed-file-path-text')].map(e =>
      (e as HTMLElement).innerHTML.replace(/\u200E/g, ''),
    );
    return found;
  }

  it('Paginates generated files', async () => {
    await simulateGeneratedFiles(1200);
    // 1200 files, but 1000 files per fetched batch of generated statuses.
    // Sorted by status, that puts 1000/3 manual files, then 1000/3 partially generated, then 1000/3 generated,
    // then the remaining 200/3 manual, 200/3 partially generated, and 200/3 generated,
    // all in pages of 500 at a time.

    // first page is manual and partial
    expectHasPartiallyGeneratedFiles();
    expectNOTHasGeneratedFiles();
    expect(getChangedFiles()).toMatchSnapshot();

    // next page has partial and generated
    goToNextPage();
    expectHasPartiallyGeneratedFiles();
    expectHasGeneratedFiles();
    expect(getChangedFiles()).toMatchSnapshot();

    // next page has remaining files from all 3 types
    goToNextPage();
    expectHasPartiallyGeneratedFiles();
    expectHasGeneratedFiles();
    expect(getChangedFiles()).toMatchSnapshot();
  });

  it('Warns about too many files to fetch all generated statuses', async () => {
    await simulateGeneratedFiles(1001);
    expect(
      screen.getByText('There are more than 1000 files, some files may appear out of order'),
    ).toBeInTheDocument();
  });
});
