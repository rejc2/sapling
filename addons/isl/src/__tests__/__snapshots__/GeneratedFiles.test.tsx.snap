// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated Files Paginates generated files 1`] = `
Array [
  "file_002.txt",
  "file_005.txt",
  "file_008.txt",
  "file_011.txt",
  "file_014.txt",
  "file_017.txt",
  "file_020.txt",
  "file_023.txt",
  "file_026.txt",
  "file_029.txt",
  "file_032.txt",
  "file_035.txt",
  "file_038.txt",
  "file_041.txt",
  "file_044.txt",
  "file_047.txt",
  "file_050.txt",
  "file_053.txt",
  "file_056.txt",
  "file_059.txt",
  "file_062.txt",
  "file_065.txt",
  "file_068.txt",
  "file_071.txt",
  "file_074.txt",
]
`;

exports[`Generated Files Paginates generated files 2`] = `
Array [
  "file_077.txt",
  "file_080.txt",
  "file_083.txt",
  "file_086.txt",
  "file_089.txt",
  "file_092.txt",
  "file_095.txt",
  "file_098.txt",
  "file_101.txt",
  "file_104.txt",
  "file_107.txt",
  "file_110.txt",
  "file_113.txt",
  "file_116.txt",
  "file_119.txt",
  "file_122.txt",
  "file_125.txt",
  "file_128.txt",
  "file_131.txt",
  "file_134.txt",
  "file_137.txt",
  "file_140.txt",
  "file_143.txt",
  "file_146.txt",
  "file_149.txt",
]
`;

exports[`Generated Files Paginates generated files 3`] = `
Array [
  "file_152.txt",
  "file_155.txt",
  "file_158.txt",
  "file_161.txt",
  "file_164.txt",
  "file_167.txt",
  "file_170.txt",
  "file_173.txt",
  "file_176.txt",
  "file_179.txt",
  "file_182.txt",
  "file_185.txt",
  "file_188.txt",
  "file_191.txt",
  "file_194.txt",
  "file_197.txt",
  "file_200.txt",
  "file_203.txt",
  "file_206.txt",
  "file_209.txt",
  "file_212.txt",
  "file_215.txt",
  "file_218.txt",
  "file_221.txt",
  "file_224.txt",
]
`;

exports[`Generated Files Paginates generated files 4`] = `
Array [
  "file_227.txt",
  "file_230.txt",
  "file_233.txt",
  "file_236.txt",
  "file_239.txt",
  "file_242.txt",
  "file_245.txt",
  "file_248.txt",
  "file_251.txt",
  "file_254.txt",
  "file_257.txt",
  "file_260.txt",
  "file_263.txt",
  "file_266.txt",
  "file_269.txt",
  "file_272.txt",
  "file_275.txt",
  "file_278.txt",
  "file_281.txt",
  "file_284.txt",
  "file_287.txt",
  "file_290.txt",
  "file_293.txt",
  "file_296.txt",
  "file_299.txt",
]
`;

exports[`Generated Files Paginates generated files 5`] = `
Array [
  "file_302.txt",
  "file_305.txt",
  "file_308.txt",
  "file_311.txt",
  "file_314.txt",
  "file_317.txt",
  "file_320.txt",
  "file_323.txt",
  "file_326.txt",
  "file_329.txt",
  "file_332.txt",
  "file_335.txt",
  "file_338.txt",
  "file_341.txt",
  "file_344.txt",
  "file_347.txt",
  "file_350.txt",
  "file_353.txt",
  "file_356.txt",
  "file_359.txt",
  "file_362.txt",
  "file_365.txt",
  "file_368.txt",
  "file_371.txt",
  "file_374.txt",
]
`;

exports[`Generated Files Paginates generated files 6`] = `
Array [
  "file_377.txt",
  "file_380.txt",
  "file_383.txt",
  "file_386.txt",
  "file_389.txt",
  "file_392.txt",
  "file_395.txt",
  "file_398.txt",
  "file_001.txt",
  "file_004.txt",
  "file_007.txt",
  "file_010.txt",
  "file_013.txt",
  "file_016.txt",
  "file_019.txt",
  "file_022.txt",
  "file_025.txt",
  "file_028.txt",
  "file_031.txt",
  "file_034.txt",
  "file_037.txt",
  "file_040.txt",
  "file_043.txt",
  "file_046.txt",
  "file_049.txt",
]
`;

exports[`Generated Files Paginates generated files 7`] = `
Array [
  "file_052.txt",
  "file_055.txt",
  "file_058.txt",
  "file_061.txt",
  "file_064.txt",
  "file_067.txt",
  "file_070.txt",
  "file_073.txt",
  "file_076.txt",
  "file_079.txt",
  "file_082.txt",
  "file_085.txt",
  "file_088.txt",
  "file_091.txt",
  "file_094.txt",
  "file_097.txt",
  "file_100.txt",
  "file_103.txt",
  "file_106.txt",
  "file_109.txt",
  "file_112.txt",
  "file_115.txt",
  "file_118.txt",
  "file_121.txt",
  "file_124.txt",
]
`;

exports[`Generated Files Paginates generated files 8`] = `
Array [
  "file_127.txt",
  "file_130.txt",
  "file_133.txt",
  "file_136.txt",
  "file_139.txt",
  "file_142.txt",
  "file_145.txt",
  "file_148.txt",
  "file_151.txt",
  "file_154.txt",
  "file_157.txt",
  "file_160.txt",
  "file_163.txt",
  "file_166.txt",
  "file_169.txt",
  "file_172.txt",
  "file_175.txt",
  "file_178.txt",
  "file_181.txt",
  "file_184.txt",
  "file_187.txt",
  "file_190.txt",
  "file_193.txt",
  "file_196.txt",
  "file_199.txt",
]
`;

exports[`Generated Files Paginates generated files 9`] = `
Array [
  "file_202.txt",
  "file_205.txt",
  "file_208.txt",
  "file_211.txt",
  "file_214.txt",
  "file_217.txt",
  "file_220.txt",
  "file_223.txt",
  "file_226.txt",
  "file_229.txt",
  "file_232.txt",
  "file_235.txt",
  "file_238.txt",
  "file_241.txt",
  "file_244.txt",
  "file_247.txt",
  "file_250.txt",
  "file_253.txt",
  "file_256.txt",
  "file_259.txt",
  "file_262.txt",
  "file_265.txt",
  "file_268.txt",
  "file_271.txt",
  "file_274.txt",
]
`;

exports[`Generated Files Paginates generated files 10`] = `
Array [
  "file_277.txt",
  "file_280.txt",
  "file_283.txt",
  "file_286.txt",
  "file_289.txt",
  "file_292.txt",
  "file_295.txt",
  "file_298.txt",
  "file_301.txt",
  "file_304.txt",
  "file_307.txt",
  "file_310.txt",
  "file_313.txt",
  "file_316.txt",
  "file_319.txt",
  "file_322.txt",
  "file_325.txt",
  "file_328.txt",
  "file_331.txt",
  "file_334.txt",
  "file_337.txt",
  "file_340.txt",
  "file_343.txt",
  "file_346.txt",
  "file_349.txt",
]
`;

exports[`Generated Files Paginates generated files 11`] = `
Array [
  "file_352.txt",
  "file_355.txt",
  "file_358.txt",
  "file_361.txt",
  "file_364.txt",
  "file_367.txt",
  "file_370.txt",
  "file_373.txt",
  "file_376.txt",
  "file_379.txt",
  "file_382.txt",
  "file_385.txt",
  "file_388.txt",
  "file_391.txt",
  "file_394.txt",
  "file_397.txt",
]
`;

exports[`Generated Files Paginates generated files 12`] = `Array []`;

exports[`Generated Files Paginates generated files 13`] = `Array []`;

exports[`Generated Files Paginates generated files 14`] = `Array []`;

exports[`Generated Files Paginates generated files 15`] = `Array []`;

exports[`Generated Files Paginates generated files 16`] = `Array []`;

exports[`Generated Files Paginates generated files 17`] = `
Array [
  "file_401.txt",
  "file_404.txt",
  "file_407.txt",
  "file_410.txt",
  "file_413.txt",
  "file_416.txt",
  "file_419.txt",
  "file_422.txt",
  "file_425.txt",
  "file_428.txt",
  "file_431.txt",
  "file_434.txt",
  "file_437.txt",
  "file_440.txt",
  "file_443.txt",
  "file_446.txt",
  "file_449.txt",
  "file_452.txt",
  "file_455.txt",
  "file_458.txt",
  "file_461.txt",
  "file_464.txt",
  "file_467.txt",
  "file_470.txt",
  "file_473.txt",
]
`;
